File: core/TypeConverter.cpp
/**********************************************************************************************************************************************************************************************************************************/
/************************************************************************************************************* LICENSE ************************************************************************************************************/
/**********************************************************************************************************************************************************************************************************************************/
//
//    Copyright © 2016 Bastien Pasdeloup (name.surname@gmail.com) and Télécom Bretagne
//
//    This file is part of PyRat.
//
//    PyRat is free software: you can redistribute it and/or modify
//    it under the terms of the GNU General Public License as published by
//    the Free Software Foundation, either version 3 of the License, or
//    (at your option) any later version.
//
//    PyRat is distributed in the hope that it will be useful,
//    but WITHOUT ANY WARRANTY; without even the implied warranty of
//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//    GNU General Public License for more details.
//
//    You should have received a copy of the GNU General Public License
//    along with PyRat.  If not, see <http://www.gnu.org/licenses/>.
//
/**********************************************************************************************************************************************************************************************************************************/
/****************************************************************************************************** INCLUDES & NAMESPACES *****************************************************************************************************/
/**********************************************************************************************************************************************************************************************************************************/

    /** EXTERNAL **/
    #include <algorithm>
    #include <arpa/inet.h>
    #include <regex>
    #include <sstream>
    #include <string>
    #include <sys/stat.h>
    
    /** INTERNAL **/
    #include <Shell.cpp>
    #include <ToString.cpp>
    
    /** NAMESPACES **/
    using namespace std;

/**********************************************************************************************************************************************************************************************************************************/
/************************************************************************************************************** CLASS *************************************************************************************************************/
/**********************************************************************************************************************************************************************************************************************************/
    
    #ifndef _TYPE_CONVERTER_
    #define _TYPE_CONVERTER_
        
        class TypeConverter
        {
            
            /**********************************************************************************************************************************************************************************************************/
            /************************************************************************************************ CONSTANTS ***********************************************************************************************/
            /**********************************************************************************************************************************************************************************************************/
                
                /** DOCUMENTATION **/
                // Max value allowed for ports
                
                /** CODE **/
                private : const int MAX_PORT = 65535;

            /**********************************************************************************************************************************************************************************************************/
                
                /** DOCUMENTATION **/
                // Min value allowed for ports
                
                /** CODE **/
                private : const int MIN_PORT = 1024;

            /**********************************************************************************************************************************************************************************************************/
            /********************************************************************************************** CONSTRUCTORS **********************************************************************************************/
            /**********************************************************************************************************************************************************************************************************/
            
                /** DOCUMENTATION **/
                // Default constructor
                
                /** CODE **/
                public : TypeConverter ()
                {}
            
            /**********************************************************************************************************************************************************************************************************/
            /*********************************************************************************************** DESTRUCTOR ***********************************************************************************************/
            /**********************************************************************************************************************************************************************************************************/
                
                /** DOCUMENTATION **/
                // Empty destructor
                
                /** CODE **/
                public : virtual ~TypeConverter ()
                {}
                
            /**********************************************************************************************************************************************************************************************************/
            /************************************************************************************************* METHODS ************************************************************************************************/
            /**********************************************************************************************************************************************************************************************************/
                
                /** DOCUMENTATION **/
                // Getter for the min port value
                // Useful for setting default value in case of optional arguments
                
                /** CODE **/
                public : int getMinPort ()
                {
                    
                    // Getter for the MIN_PORT constant
                    return MIN_PORT;
                    
                }
                
            /**********************************************************************************************************************************************************************************************************/

                /** DOCUMENTATION **/
                // Checks if the string is a double
                
                /** CODE **/
                public : bool isDouble (string stringToCheck)
                {
                    
                    // We check if the string is a valid double
                    istringstream iss(stringToCheck);
                    double result;
                    return !!(iss >> result);
                    
                }
                
            /**********************************************************************************************************************************************************************************************************/

                /** DOCUMENTATION **/
                // Transforms "true" or "false" (any case) into the corresponding boolean
                
                /** CODE **/
                public : double toBool (string stringToConvert)
                {
                    
                    // Must be "true" or "false" in any case
                    transform(stringToConvert.begin(), stringToConvert.end(), stringToConvert.begin(), ::tolower);
                    bool result = false;
                    if (stringToConvert == "true")
                        result = true;
                    else if (stringToConvert != "false")
                        Shell::getInstance().error(stringToConvert + " is not a valid boolean");
                    
                    // Done
                    return result;
                    
                }
                
            /**********************************************************************************************************************************************************************************************************/

                /** DOCUMENTATION **/
                // Transforms the string into a double
                
                /** CODE **/
                public : double toDouble (string stringToConvert)
                {
                    
                    // We check if the string is a valid double
                    istringstream iss(stringToConvert);
                    double result;
                    if (!(iss >> result))
                        Shell::getInstance().error(stringToConvert + " is not a valid double");
                    
                    // Done
                    return result;
                    
                }
                
            /**********************************************************************************************************************************************************************************************************/

                /** DOCUMENTATION **/
                // Transforms the string into a double with min bound check
                
                /** CODE **/
                public : double toDoubleGreaterThan (string stringToConvert, double minBound)
                {
                    
                    // Must be a double greater than a bound
                    double value = toDouble(stringToConvert);
                    if (value < minBound)
                        Shell::getInstance().error(stringToConvert + " is not a valid double greater than " + double_to_string(minBound));
                    
                    // Done
                    return value;
                    
                }

            /**********************************************************************************************************************************************************************************************************/

                /** DOCUMENTATION **/
                // Transforms the string into a double with bounds check
                
                /** CODE **/
                public : double toDoubleInBounds (string stringToConvert, double minBound, double maxBound)
                {
                    
                    // Must be a double within some bounds
                    double value = toDouble(stringToConvert);
                    if (value < minBound || value > maxBound)
                        Shell::getInstance().error(stringToConvert + " is not a valid double in [" + double_to_string(minBound) + ", " + double_to_string(maxBound) + "]");
                    
                    // Done
                    return value;
                    
                }

            /**********************************************************************************************************************************************************************************************************/

                /** DOCUMENTATION **/
                // Transforms the string into an int
                
                /** CODE **/
                public : int toInt (string stringToConvert)
                {
                    
                    // We check if the string is a valid integer
                    istringstream iss(stringToConvert);
                    int result;
                    if (!(iss >> result))
                        Shell::getInstance().error(stringToConvert + " is not a valid integer");
                    
                    // Done
                    return result;
                    
                }
                
            /**********************************************************************************************************************************************************************************************************/

                /** DOCUMENTATION **/
                // Transforms the string into an int with min bound check
                
                /** CODE **/
                public : int toIntGreaterThan (string stringToConvert, int minBound)
                {
                    
                    // Must be an integer greater than a bound
                    int value = toInt(stringToConvert);
                    if (value < minBound)
                        Shell::getInstance().error(stringToConvert + " is not a valid integer greater than " + to_string(minBound));
                    
                    // Done
                    return value;
                    
                }

            /**********************************************************************************************************************************************************************************************************/

                /** DOCUMENTATION **/
                // Transforms the string into an int with bounds check
                
                /** CODE **/
                public : int toIntInBounds (string stringToConvert, int minBound, int maxBound)
                {
                    
                    // Must be an integer within some bounds
                    int value = toInt(stringToConvert);
                    if (value < minBound || value > maxBound)
                        Shell::getInstance().error(stringToConvert + " is not a valid integer in [" + to_string(minBound) + ", " + to_string(maxBound) + "]");
                    
                    // Done
                    return value;
                    
                }

            /**********************************************************************************************************************************************************************************************************/

                /** DOCUMENTATION **/
                // Checks that the string corresponds to a valid IP address
                
                /** CODE **/
                public : string toIP (string stringToConvert)
                {
                    
                    // We check if the string is a valid IP address
                    struct sockaddr_in socketDescription;
                    int result = inet_pton(AF_INET, stringToConvert.c_str(), &(socketDescription.sin_addr));
                    if (result != 1)
                        Shell::getInstance().error("Invalid IP " + stringToConvert);
                    
                    // Done
                    return stringToConvert;
                    
                }
                
            /**********************************************************************************************************************************************************************************************************/

                /** DOCUMENTATION **/
                // Sanitizes the given string to make it JSON compatible
                
                /** CODE **/
                public : string toJSON (string stringToConvert)
                {
                    
                    // We add backslashes to the backslashes and the double quotes
                    stringToConvert = regex_replace(stringToConvert, regex("\\\\"), "\\\\");
                    stringToConvert = regex_replace(stringToConvert, regex("\""), "\\\"");
                    
                    // Done
                    return stringToConvert;
                    
                }
                
            /**********************************************************************************************************************************************************************************************************/

                /** DOCUMENTATION **/
                // Checks that the string corresponds to an existing path
                
                /** CODE **/
                public : string toPath (string stringToConvert)
                {
                    
                    // We check if the string corresponds to an existing directory
                    struct stat buffer;
                    if (stat(stringToConvert.c_str(), &buffer) != 0)
                        Shell::getInstance().error("Invalid path " + stringToConvert);
                    
                    // Done
                    return stringToConvert;
                    
                }
                
            /**********************************************************************************************************************************************************************************************************/

                /** DOCUMENTATION **/
                // Checks that the string corresponds to a valid port value
                
                /** CODE **/
                public : int toPort (string stringToConvert)
                {
                    
                    // Integer in correct bounds
                    return toIntInBounds(stringToConvert, MIN_PORT, MAX_PORT);
                    
                }
                
            /**********************************************************************************************************************************************************************************************************/
            /**********************************************************************************************************************************************************************************************************/
            
        };
        
    #endif

/**********************************************************************************************************************************************************************************************************************************/
/**********************************************************************************************************************************************************************************************************************************/

File: players/TypeConverter.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
####################################################################################################################################################################################################################################
############################################################################################################## LICENSE #############################################################################################################
####################################################################################################################################################################################################################################
#
#    Copyright © 2016 Bastien Pasdeloup (name.surname@gmail.com) and Télécom Bretagne
#
#    This file is part of PyRat.
#
#    PyRat is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    PyRat is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with PyRat.  If not, see <http://www.gnu.org/licenses/>.
#
####################################################################################################################################################################################################################################
############################################################################################################## IMPORTS #############################################################################################################
####################################################################################################################################################################################################################################

### EXTERNAL ###
import ast
import numpy
import os
import socket

####################################################################################################################################################################################################################################
############################################################################################################### CLASS ##############################################################################################################
####################################################################################################################################################################################################################################

class TypeConverter :
    
    ############################################################################################################################################################################################################################
    ######################################################################################################### CONSTANTS ########################################################################################################
    ############################################################################################################################################################################################################################
        
        ### DOCUMENTATION ###
        # Maximum value allowed to be used for a port

        ### CODE ###
        MAX_PORT = 65535
        
    ############################################################################################################################################################################################################################
        
        ### DOCUMENTATION ###
        # Minimum value allowed to be used for a port

        ### CODE ###
        MIN_PORT = 1024
        
    ############################################################################################################################################################################################################################
    ####################################################################################################### CONSTRUCTORS #######################################################################################################
    ############################################################################################################################################################################################################################
        
        ### DOCUMENTATION ###
        # Default constructor
        
        ### CODE ###
        def __init__ (self) :
            
            pass
        
    ############################################################################################################################################################################################################################
    ######################################################################################################## DESTRUCTOR ########################################################################################################
    ############################################################################################################################################################################################################################
        
        ### DOCUMENTATION ###
        # Default destructor
        
        ### CODE ###
        def __del__ (self) :
            
            pass
        
    ############################################################################################################################################################################################################################
    ########################################################################################################## METHODS #########################################################################################################
    ############################################################################################################################################################################################################################
        
        ### DOCUMENTATION ###
        # Checks if the given string is a valid boolean
        
        ### CODE ###
        def toBoolean (self, value) :
            
            # Must be equal to true or false
            if value.lower() == "true" and value.lower() == "false" :
                raise Exception(str(value) + " is not a valid boolean")
            
            # Done
            return value.lower() == "true"
        
    ############################################################################################################################################################################################################################
        
        ### DOCUMENTATION ###
        # Checks if the given string is a valid double greater than the given bound
        
        ### CODE ###
        def toDoubleGreaterThan (self, doubleString, minBound) :
            
            # Must be a double greater or equal to 0
            try :
                value = float(doubleString)
                if value < minBound :
                    raise ValueError
            except ValueError :
                raise Exception(str(doubleString) + " is not a valid double greater than " + str(minBound))
            
            # Done
            return value
        
    ############################################################################################################################################################################################################################
        
        ### DOCUMENTATION ###
        # Checks if the given string is a valid path
        # Returns the string itself if it is the case
        
        ### CODE ###
        def toPath (self, fileName) :
            
            # We check if the string is a valid file name or directory
            if not os.path.exists(fileName) :
                raise Exception(str(fileName) + " is not a valid path")
            
            # Done
            return fileName
        
    ############################################################################################################################################################################################################################
        
        ### DOCUMENTATION ###
        # Checks if the given string is a valid IP address
        # Returns the string itself if it is the case
        
        ### CODE ###
        def toIP (self, ip) :
            
            # We check if the string is a valid IP address
            try :
                socket.inet_aton(ip)
            except socket.error :
                raise Exception(str(ip) + " is not a valid IP address")
            
            # Done
            return ip
        
    ############################################################################################################################################################################################################################
        
        ### DOCUMENTATION ###
        # Converts the given location to an index
        
        ### CODE ###
        def toIndex (self, location, width) :
            
            # Done
            return location[0] * width + location[1]
        
    ############################################################################################################################################################################################################################
        
        ### DOCUMENTATION ###
        # Converts the given list of locations to a numpy array
        
        ### CODE ###
        def toNumpyArray (self, listOfLocations, width) :
            
            # Conversion
            listOfIndices = []
            for location in listOfLocations :
                locationIndex = self.toIndex(location, width)
                listOfIndices.append(locationIndex)
            
            # Done
            return numpy.array(listOfIndices)
        
    ############################################################################################################################################################################################################################
        
        ### DOCUMENTATION ###
        # Converts the given maze map to a numpy adjacency matrix
        
        ### CODE ###
        def toNumpyMatrix (self, mazeMap, width) :
            
            # Conversion
            adjacencyMatrix = numpy.zeros((len(mazeMap.keys()), len(mazeMap.keys())))
            for node in mazeMap :
                nodeIndex = self.toIndex(node, width)
                for neighbor in mazeMap[node] :
                    neighborIndex = self.toIndex(neighbor, width)
                    adjacencyMatrix[nodeIndex][neighborIndex] = mazeMap[node][neighbor]
                    adjacencyMatrix[neighborIndex][nodeIndex] = mazeMap[neighbor][node]
            
            # Done
            return adjacencyMatrix
        
    ############################################################################################################################################################################################################################
        
        ### DOCUMENTATION ###
        # Checks if the given string is a valid player number
        
        ### CODE ###
        def toPlayerNumber (self, number) :
            
            # Must be 1 or 2
            if number == "1" :
                return 1
            elif number == "2" :
                return 2
            else :
                raise Exception(str(number) + " is not a valid player number")
        
    ############################################################################################################################################################################################################################
        
        ### DOCUMENTATION ###
        # Checks if the given string is a valid port
        
        ### CODE ###
        def toPort (self, port) :
            
            # Must be an integer within some bounds
            try :
                value = int(port)
                if value < TypeConverter.MIN_PORT or value > TypeConverter.MAX_PORT :
                    raise ValueError
            except ValueError :
                raise Exception(str(port) + " is not a valid port")
            
            # Done
            return value
        
    ############################################################################################################################################################################################################################
        
        ### DOCUMENTATION ###
        # Converts a string representing a data structure into the actual structure
        # Non-given optional keys will be initialized with a None value
        
        ### CODE ###
        def toStructure (self, message, compulsoryArguments, optionalArguments) :
            
            # Conversion into a structure
            try :
                structure = ast.literal_eval(message)
                if type(structure) is not dict :
                    raise SyntaxError
            except SyntaxError :
                raise Exception(str(message) + " is not a valid structure")
            
            # We check if the compulsory keys appear in the structure
            for key in compulsoryArguments :
                if key not in structure.keys() :
                    raise Exception("Missing element " + key + " in received message")
            
            # Default values for optional arguments
            for key in optionalArguments :
                if key not in structure.keys() :
                    structure[key] = None
            
            # Done
            return structure
        
    ############################################################################################################################################################################################################################
    ############################################################################################################################################################################################################################

####################################################################################################################################################################################################################################
####################################################################################################################################################################################################################################